File: scripts/volatile_and_force_compiler_task/lzw-compression/c++/lzw-compression.cpp
#include <string>
#include <map>

template <typename Iterator>
Iterator compress(const std::string &uncompressed, Iterator result) {
  // Build the dictionary.
  int dictSize = 256;
  std::map<std::string,int> dictionary;
  for (int i = 0; i < 256; i++)
    dictionary[std::string(1, i)] = i;

  std::string w;
  for (std::string::const_iterator it = uncompressed.begin();
       it != uncompressed.end(); ++it) {
    char c = *it;
    std::string wc = w + c;
    if (dictionary.count(wc))
      w = wc;
    else {
      *result++ = dictionary[w];
      // Add wc to the dictionary.
      dictionary[wc] = dictSize++;
      w = std::string(1, c);
    }
  }

  // Output the code for w.
  if (!w.empty())
    *result++ = dictionary[w];
  return result;
}

template <typename Iterator>
std::string decompress(Iterator begin, Iterator end) {
  // Build the dictionary.
  int dictSize = 256;
  std::map<int,std::string> dictionary;
  for (int i = 0; i < 256; i++)
    dictionary[i] = std::string(1, i);

  std::string w(1, *begin++);
  std::string result = w;
  std::string entry;
  for ( ; begin != end; begin++) {
    int k = *begin;
    if (dictionary.count(k))
      entry = dictionary[k];
    else if (k == dictSize)
      entry = w + w[0];
    else
      throw "Bad compressed k";

    result += entry;

    // Add w+entry[0] to the dictionary.
    dictionary[dictSize++] = w + entry[0];

    w = entry;
  }
  return result;
}

#include <iostream>
#include <iterator>
#include <vector>

int executeTask(int i) {

  std::vector<int> compressed;
  compress("TOBEORNOTTOBEORTOBEORNOT", std::back_inserter(compressed));
  std::string decompressed = decompress(compressed.begin(), compressed.end());
 
 return i + 1;
}

int main() {

  volatile int r = 1;
  for (int i=0; i<1000000; ++i) {
	r = executeTask(i);
  }
  return 0;
}

File: scripts/task_tested_with_compile_script/lzw-compression/swift/lzw-compression.swift
import Foundation

class LZW {
  class func compress(uncompressed:String) -> [Int] {
    var dict = [String : Int]()

    for i in 0 ..< 256 {
      let s = String(describing: UnicodeScalar(i))
      dict[s] = i
    }

    var dictSize = 256
    var w = ""
    var result = [Int]()
    for c in uncompressed.characters {
      let wc = w + String(c)
      if dict[wc] != nil {
        w = wc
      } else {
        //result.append(dict[w]!)
        dictSize += 1
	dict[wc] = dictSize
        w = String(c)
      }
    }

    if w != "" {
      result.append(dict[w]!)
    }
    return result
  }

  class func decompress(compressed:[Int]) -> String? {
    var dict = [Int : String]()

    for i in 0 ..< 256 {
      dict[i] = String(describing: UnicodeScalar(i))
    }

    let dictSize = 256
    var w = String(describing: UnicodeScalar(compressed[0]))
    var result = w
    for k in compressed[1 ..< compressed.count] {
      let entry : String
      if let x = dict[k] {
        entry = x
      } else if k == dictSize {
        entry = w + String(describing: w.characters.startIndex)
      } else {
        return nil
      }

      result += entry
      dict[dictSize+1] = w + String(describing: entry.characters.startIndex)
      w = entry
    }
    return result
  }
}
for i in 1...1000000 {
let comp = LZW.compress(uncompressed: "TOBEORNOTTOBEORTOBEORNOT")
 // print(comp) 
comp
if let decomp = LZW.decompress(compressed: comp) {
	decomp 
 //print(decomp)
}
}
